mllib/src/main/scala/org/apache/spark/ml/tree/impl/LocalDecisionTree.scala

```diff
@@ -0,0 +1,250 @@
+    }
+
+    // Prepare column store.
+    //   Note: rowToColumnStoreDense checks to make sure numRows < Int.MaxValue.
+    val colStoreInit: Array[Array[Int]]
+    = LocalDecisionTreeUtils.rowToColumnStoreDense(input.map(_.binnedFeatures))
+    val labels = input.map(_.label)
+
+    // Fit a regression model on the dataset, throwing an error if metadata indicates that
+    // we should train a classifier.
+    // TODO: Add support for training classifiers
+    if (metadata.numClasses > 1 && metadata.numClasses <= 32) {
+      throw new UnsupportedOperationException("Local training of a decision tree classifier is " +
+        "unsupported; currently, only regression is supported")
+    } else {
+      trainRegressor(node, colStoreInit, instanceWeights, labels, metadata, splits)
+    }
+  }
+
+  /**
+   * Locally fits a decision tree regressor.
+   * TODO(smurching): Logic for fitting a classifier & regressor is the same; only difference
+   * is impurity metric. Use the same logic for fitting a classifier.
+   *
+   * @param rootNode Node to use as root of the tree fit on the passed-in dataset
+   * @param colStoreInit Array of columns of training data
+   * @param instanceWeights Array of weights for each training example
+   * @param metadata learning and dataset metadata for DecisionTree
+   * @param splits splits(i) = Array of possible splits for feature i
+   * @return LeafNode or InternalNode representation of rootNode
+   */
+  private[ml] def trainRegressor(
+      rootNode: LearningNode,
+      colStoreInit: Array[Array[Int]],
+      instanceWeights: Array[Double],
+      labels: Array[Double],
+      metadata: DecisionTreeMetadata,
+      splits: Array[Array[Split]]): Node = {
+
+    // Sort each column by decision tree node.
+    val colStore: Array[FeatureVector] = colStoreInit.zipWithIndex.map { case (col, featureIndex) =>
+      val featureArity: Int = metadata.featureArity.getOrElse(featureIndex, 0)
+      FeatureVector(featureIndex, featureArity, col)
+    }
+
+    val numRows = colStore.headOption match {
+      case None => 0
+      case Some(column) => column.values.length
+    }
+
+    // Create a new PartitionInfo describing the status of our partially-trained subtree
+    // at each iteration of training
+    var trainingInfo: TrainingInfo = TrainingInfo(colStore, instanceWeights,
+      nodeOffsets = Array[(Int, Int)]((0, numRows)), activeNodes = Array(rootNode))
+
+    // Iteratively learn, one level of the tree at a time.
+    // Note: We do not use node IDs.
+    var currentLevel = 0
+    var doneLearning = false
+
+    while (currentLevel < metadata.maxDepth && !doneLearning) {
+      // Splits each active node if possible, returning an array of new active nodes
+      val activeNodes: Array[LearningNode] =
+        computeBestSplits(trainingInfo, labels, metadata, splits)
+      // Filter active node periphery by impurity.
+      val estimatedRemainingActive = activeNodes.count(_.stats.impurity > 0.0)
```

Comment:
Agreed on using `isLeaf` instead of checking for positive impurity, thanks for the suggestion.

AFAICT at this point in the code `activeNodes` actually does refer to the nodes in the current level; the children of nodes in `activeNodes` are the nodes in the next level, and are returned by `computeBestSplits`. I forgot to include the return type of `computeBestSplit` in its method signature, which probably made this more confusing - my mistake.

Reply:
Yes. Sorry for confusing you. The change what I said was changing to:
```
val nextLevelNodes: Array[LearningNode] =
        computeBestSplits(trainingInfo, labels, metadata, splits)
```
Does it look more reasonable ?
And change the member name in `trainingInfo`:
`TrainingInfo.activeNodes` ==> `TrainingInfo.currentLevelNodes`

Reply:
Gotcha agreed on the naming change, how about `currentLevelActiveNodes`? Since only the non-leaf nodes from the current level are included.

Reply:
Wait... I check the code here: `trainingInfo = trainingInfo.update(splits, activeNodes)` So it seems you do not filter out the leaf node from the "activeNodes"(which is actually the `nextLevelNode` I mentioned above).
So I think `TrainingInfo.activeNodes` is still possible to contains leaf node.

Reply:
Oh true -- I'll doc for `currentLevelActiveNodes` to say:
```
 * @param currentLevelActiveNodes  Nodes which are active (could still be split).
 *                                 Inactive nodes are known to be leaves in the final tree.
```